### ISSUE TITLE
Add SIMD optimisations on ARM64 on progressive JPEG encoding / decoding

### ISSUE THREAD

**Issue by andob** [libjpeg-turbo/libjpeg-turbo#625]
Context: inside an Android application, I am trying to decode some non-progressive image files and then re-encode them as progressive JPEG files. I noticed that the library is very fast on encoding standard JPEGs, but not so fast on encoding progressive JPEGs.

That is because SIMD optimisations are not implmented on ARM. From your website: "libjpeg-turbo is a JPEG image codec that uses SIMD instructions (MMX, SSE2, AVX2, Neon, AltiVec) to accelerate baseline JPEG compression and decompression on x86, x86-64, Arm, and PowerPC systems, as well as **progressive JPEG compression on x86 and x86-64 systems.**"

I also ran a naive benchmark:
| Device            | CPU                  | Progressive JPEG | Standard JPEG | JPEG source |
|-------------------|----------------------|------------------|---------------|---------------|
| Pixel 5           | arm64-v8a (64 bit)   | 3.1s             | 200ms         | phone's camera image ~2MB |
| Motorola Moto G22 | arm64-v8a (64 bit)   | 5.8s             | 800ms         | phone's camera image ~2MB |
| Samsung XCover 4  | armeabi-v7a (32 bit) | 12s              | 1.5s          | phone's camera image ~2MB |
| MAC mini M1   | M1 (64 bit) | 11s | 3s | very large image ~140MB |

Why there are no SIMD compression / decompression optimisations on ARM / progressive JPEG?

Nowadays, most tablets and phones do have ARM V8 (64 bit) processors. Would be great SIMD optimisations to be implemented on ARM V8. No need to struggle implementing them on ARM V7 (32 bit).

Also, with the advent of Apple's M1, ARM processors become increasingly more popular on desktop PCs and laptops, since these processors are much more energy efficient than the classic CISC processors.

Is there any plan to implement SIMD optimisations on ARM64 on progressive JPEG encoding and decoding?

**Comment by dcommander** [comment]
There actually are Arm Neon SIMD extensions for progressive Huffman encoding in libjpeg-turbo 2.1.x.  The libjpeg-turbo README file correctly reflects that, but I neglected to make the same update to the web site.  That has been corrected.

The Neon implementation of progressive Huffman encoding does provide a significant overall speedup ("overall speedup" = speedup of the entire compression operation) relative to the scalar version of the same algorithm, but that overall speedup is in the ballpark of 40-45%, which is significantly less than the 80-90% overall speedup that can be achieved with the x86-64 implementation of progressive Huffman encoding.  @jwright-arm and @mayeut may be able to provide further insight.  @mayeut wrote the Neon assembly implementation of progressive Huffman encoding, and @jwright-arm ported that code to Neon intrinsics.  (The intrinsics implementation is the code that libjpeg-turbo 2.1.x actually uses.)  @mayeut's initial assembly implementation provided about a 30-40% overall speedup, and @jwright-arm's intrinsics implementation improved the overall performance by about 5-10% relative to that.

Two things to bear in mind:
1. How you benchmark the code matters a lot.  Many people will naively run `cjpeg` as a compression benchmark, but because it interleaves compute and I/O, it is a poor benchmark.  `tjbench {PPM-or-BMP-file} {quality-level} -rgb -quiet -progressive -nowrite` is what I personally use to benchmark progressive JPEG compression.  All of that aside, my own testing shows about a 3-5x disparity in performance between baseline and progressive compression on A53 and A72 cores, so your large image test is in the same ballpark.  You can probably improve your results for the small images by benchmarking using tjbench, but you likely will still observe a 3-5x disparity between baseline and progressive compression.
2. Progressive JPEG compression becomes more computationally difficult the more scans there are.  The default scan script used by `jpeg_simple_progression()` (which is also used when you enable progressive JPEG compression in the TurboJPEG API) is not the best for all use cases.  You can probably get better performance by using fewer scans. However, specifying custom scan scripts is currently only a function of the libjpeg API.  (It will be added to the TurboJPEG API in libjpeg-turbo 2.2.)  All of that aside, the performance claims I made above were from using the default scan script, so regardless of the scan script, the Arm SIMD implementation of progressive Huffman encoding still won't provide the same speedup as the x86-64 implementation.

**Comment by andob** [comment]
thank you for your reply!

something is wrong, somewhere.

Tomorrow I will run tjbench with and without -optimize flag, on all 3 mobile devices. I will come back with the results.

**Comment by dcommander** [comment]
Clarifications/corrections to my previous comment:

Comparing the relative performance between baseline and progressive JPEG compression doesn't really tell the whole story.  I was cribbing data from various benchmark spreadsheets when I made the comment above.  The 80-90% overall speedup for progressive compression on x86-64, relative to baseline compression, was from a machine that doesn't support AVX2 extensions.  It's more illustrative to look at the absolute numbers on a more modern CPU.  Thus, I re-ran the benchmark just now on my Mac with a dual-core 3 GHz i7, using the five test images described [here](https://libjpeg-turbo.org/About/Performance).  The performance numbers below reflect an average across all five images and four subsampling types (grayscale, 4:2:0, 4:2:2, and 4:4:4.)

| CPU | Progressive JPEG Compression Performance in Mpixels/sec, Scalar Progressive Huffman Encoding |  Progressive JPEG Compression Performance in Mpixels/sec, SSE2 Progressive Huffman Encoding (speedup relative to scalar progressive Huffman encoding) |  Baseline JPEG Compression Performance in Mpixels/sec (speedup relative to progressive + SSE2 progressive Huffman encoding) |
|---|---|---|---|
| 3 GHz Dual-Core Intel Core i7 | 27.4 | 40.5 (+47.8%) | 251.6 (+6.2x) |

Now, the same numbers from my study of the Neon SIMD extensions prior to the release of libjpeg-turbo 2.1, this time on a Cortex-A72 core:

| CPU | Progressive JPEG Compression Performance in Mpixels/sec, Scalar Progressive Huffman Encoding |  Progressive JPEG Compression Performance in Mpixels/sec, Neon Progressive Huffman Encoding (speedup relative to scalar progressive Huffman encoding |  Baseline JPEG Compression Performance in Mpixels/sec (speedup relative to progressive + Neon progressive Huffman encoding) |
|---|---|---|---|
| 1.8 GHz Dual-Core Arm Cortex-A72 | 11.5 | 15.0 (+30.4%) | 73.7 (+4.9x) |

In other words, the baseline vs. progressive disparity isn't as different between x86-64 and Arm as I previously thought.  To be extra clear, progressive JPEG will *always* be *considerably* slower than baseline JPEG, regardless of the level of SIMD acceleration.  You can close that gap by reducing the number of scans, as I described above.

**Comment by andob** [comment]
I ran the benchmark. This is surprising, but jpeg-turbo SIMD compression is actually slower than scalar compression. At least that's how it is on my devices:

| Device           | SIMD baseline compression (Mpix/sec) | SIMD progressive compression (Mpix/sec) | Scalar baseline compression (Mpix/sec) | Scalar progressive compression (Mpix/sec) |
|------------------|--------------------------------------|-----------------------------------------|----------------------------------------|-------------------------------------------|
| Google Pixel 5   | 1.981                                | 1.989                                   | 1.979                                  | 1.987                                     |
| Samsung Xcover 4 | 0.6512                               | 0.4855                                  | 3.6                                    | 3.606                                     |
| Motorola G22     | 1.332                                | 1.331                                   | 1.625                                  | 1.621                                     |

I used nightshot_iso_100.ppm test file from [here](http://imagecompression.info/test_images/), quality level 85. Speed values are copied from RGB  (TD)  4:4:4 row, Comp Perf column. [Here](https://github.com/andob/libjpeg-turbo-android-bindings/tree/tjbench/results) you can find raw benchmark output files.

This was also visually confirmed with my initial code that takes a picture with phone's camera, then runs transjpeg on that picture. While Google Pixel gets about the same speed scalar vs simd, on Motorola scalar is slightly better, on Samsung scalar is way better.

Hardware specs: Phone / Chipset / CPU:
- Google Pixel 5 -> Qualcomm SM7250 Snapdragon 765G 5G (7 nm) Chipset, Octa-core (1x2.4 GHz Kryo 475 Prime & 1x2.2 GHz Kryo 475 Gold & 6x1.8 GHz Kryo 475 Silver) [source](https://www.gsmarena.com/google_pixel_5-10386.php)
- Motorola Moto G22 -> Mediatek MT6765V/CB Helio G37 (12 nm) Chipset, Octa-core (4x2.3 GHz Cortex-A53 & 4x1.8 GHz Cortex-A53) [source](https://www.gsmarena.com/motorola_moto_g22-11365.php)
- Samsung XCover 4 -> Exynos 7570 Quad (14 nm) Chipset, Quad-core 1.4 GHz Cortex-A53 [source](https://www.gsmarena.com/samsung_galaxy_xcover_4-8577.php) - this device, in theory run 64bit software, it actually runs 32bit software. ``android.os.Build.SUPPORTED_ABIS`` returns ``armeabi-v7a``, not ``arm64-v8a``.

Possible reasons:
1. Something wrong with the simd algorithms? Maybe they run great on some hardware, but they are not optimised on other hardware.
2. Something wrong with the compiler? Maybe Android NDK doesn't compile well enough those SIMD instructions.

**Comment by andob** [comment]
I also did averages on Comp Perf columns, instead of taking just the 4:4:4 cells. Results look like this:

| Device           | SIMD baseline compression (Mpix/sec) | SIMD progressive compression (Mpix/sec) | Scalar baseline compression (Mpix/sec) | Scalar progressive compression (Mpix/sec) |
|------------------|--------------------------------------|-----------------------------------------|----------------------------------------|-------------------------------------------|
| Google Pixel 5   | 3.076                                | 3.081                                   | 3.083                                  | 3.092                                     |
| Samsung Xcover 4 | 0.971                                | 0.768                                   | 5.572                                  | 5.575                                     |
| Motorola G22     | 2.076                                | 2.066                                   | 2.532                                  | 2.52                                      |

Scalar is still faster than SIMD.

**Comment by dcommander** [comment]
That's really odd that baseline and progressive perform almost the same.  That shouldn't ever be the case, and it makes me suspicious that something still isn't right.  Baseline has just a single scan, whereas the default progressive mode in libjpeg-turbo has 10 scans, so it will always be significantly more computationally complex.

**Comment by dcommander** [comment]
In terms of performance, one thing you can also try is to turn off the `NEON_INTRINSICS` CMake variable.  The build system tries to detect whether the compiler has an optimal implementation of Neon intrinsics.  When using Clang or GCC 12+, it should enable `NEON_INTRINSICS` by default, but when using older GCC versions, it should disable `NEON_INTRINSICS` by default.  If `NEON_INTRINSICS` is enabled, then the full intrinsics implementation of the Neon SIMD extensions will be used.  If it is disabled, then some algorithms will use the older assembly (GAS) implementation from libjpeg-turbo 2.0.x and prior.  It would be curious to see whether that affects performance on older hardware.  Also, benchmarking Android phones is challenging, because most of them have both in-order and out-of-order cores, and most of them have CPU speed scaling.  In order to get accurate results, you have to disable CPU speed scaling, and you have to use `taskset` to specify a core, or else the operating system will choose one for you, and you won't know whether it will choose an in-order or out-of-order core.

**Comment by andob** [comment]
I think I have a theory about the Samsung device. According to GSMArena, it has a 64 bit A53 processor. But the operating system is a 32 bit operating system. Probably guys at samsung installed 32 bit android to trick users that the device is not actually that laggy (still, it is). Automatically, 32 bit binaries are used. Even if the 32 binary contains SIMD instructions, maybe there's a missing link somewhere, thus in the end SIMD is not actually getting used by the CPU, on this device.

As of SIMD baseline vs SIMD progressive gap.. I'm not running tjbench as a standalone executable. Instead, I did wrote JNI bindings, I'm calling tjbench's main function from java. I do two calls: one with -progressive and one without. Both calls are on the same process, the same thread, one after another. So it doesn't matter much if the OS decides to allocate the thread to a energy efficient core or a power core. Both tjbench calls will be on the same core, since they are on the same thread.

Why the resulting time is very similar in both cases is, indeed, a mistery.

Maybe you could try to tjbench on a A53 CPU and on a Qualcomm Kryo 475 CPU.

**Comment by dcommander** [comment]
The only Arm system I have available for testing is a Rock 960, based on an RK3399 SoC with dual Cortex-A72 cores and quad Cortex-A53 cores.  The A72's are definitely much faster, but they also tend to have a lot more performance variation, so they're hard to benchmark (even when CPU scaling is disabled.)

Even in the benchmark scenario you described, CPU scaling could kick in at any time during the run, so it's possible that the CPU is running faster earlier in the run.

**Comment by dcommander** [comment]
I also have a hard time imagining why the scalar and SIMD performance would be so close on the Pixel 5.  When you say "scalar", did you disable all of the SIMD extensions or just the progressive Huffman encoding SIMD extension?  You should see 2-4x faster performance with SIMD when compressing baseline images.

**Comment by andob** [comment]
by SIMD I mean C_ARITH_CODING_SUPPORTED 1
D_ARITH_CODING_SUPPORTED 1
WITH_SIMD 1
REQUIRE_SIMD 1
NEON_INTRINSICS 1
-DANDROID_ARM_NEON=ON

by scalar I mean C_ARITH_CODING_SUPPORTED 1
D_ARITH_CODING_SUPPORTED 1
WITH_SIMD 0
REQUIRE_SIMD 0
NEON_INTRINSICS 0
-DANDROID_ARM_NEON=ON commented out

I don't know how to disable / enable just the simd progressive encoding extension. When I do with_simd and neon_intrinsics, it is enabled by default, right? Is there a flag to enable / disable just the progressive simd encoding part?

**Comment by dcommander** [comment]
There is no flag to disable just progressive Huffman encoding. I had to modify jsimd.c in order to accomplish that. But you did correctly disable and enable all SIMD extensions, so you should see some difference relative to the scalar performance, and you should see some difference between baseline and progressive.

**Comment by andob** [comment]
yeah, it was CPU scaling kicking in, affecting the results. Now I did run all 4 tests separately:

| Device           | SIMD baseline compression (Mpix/sec) | SIMD progressive compression (Mpix/sec) | Scalar baseline compression (Mpix/sec) | Scalar progressive compression (Mpix/sec) |
|------------------|--------------------------------------|-----------------------------------------|----------------------------------------|-------------------------------------------|
| Google Pixel 5   | 9.480                                | 3.086                                   | 16.972                                 | 5.561                                     |

Good to know, would be best for application developers to interleave decompress/compress operations with less CPU-intensive tasks, such as network I/O, awaiting server response --- rather than queueing decompress/compress operations one after another.

**Comment by dcommander** [comment]
That's good to know.  `NEON_INTRINSICS` should be enabled automatically as long as you are using Clang, but it doesn't hurt to force it (just be sure not to force it if you are using GCC.)

Those performance numbers are in line with what I would expect.
